ipa-core/src/protocol/hybrid/agg.rs

```diff
@@ -0,0 +1,250 @@
+                }
+                MatchEntry::Pair(_, _) => {
+                    reports_by_matchkey.insert(match_key, MatchEntry::MoreThanTwo);
+                }
+                MatchEntry::MoreThanTwo => (),
+            },
+            None => {
+                reports_by_matchkey.insert(match_key, MatchEntry::Single(report.into()));
+            }
+        };
+    }
+
+    // we only keep the reports from match_keys that provided exactly 2 reports
+    reports_by_matchkey
+        .into_values()
+        .filter_map(|match_entry| match match_entry {
+            MatchEntry::Pair(r1, r2) => Some([r1, r2]),
+            _ => None,
+        })
+        .collect::<Vec<_>>()
+}
+
+/// This protocol is used to aggregate `PRFHybridReports` and returns `AggregateableHybridReports`.
+/// It groups all the reports by the PRF of the `match_key`, finds all reports from `match_keys`
+/// with that provided exactly 2 reports, then adds those 2 reports.
+pub async fn aggregate_reports<BK, V, C>(
+    ctx: C,
+    reports: Vec<PrfHybridReport<BK, V>>,
+) -> Result<Vec<AggregateableHybridReport<BK, V>>, Error>
+where
+    C: UpgradableContext + ShardedContext,
+    BK: BooleanArray,
+    V: BooleanArray,
+    Replicated<Boolean>: BooleanProtocols<DZKPUpgraded<C>>,
+{
+    let report_pairs = group_report_pairs(reports);
+
+    let chunk_size: usize = TARGET_PROOF_SIZE / (BK::BITS as usize + V::BITS as usize);
+
+    let ctx = ctx.set_total_records(TotalRecords::specified(report_pairs.len())?);
+
+    let dzkp_validator = ctx.dzkp_validator(
+        MaliciousProtocolSteps {
+            protocol: &HybridStep::GroupBySum,
+            validate: &HybridStep::GroupBySumValidate,
+        },
+        chunk_size.next_power_of_two(),
+    );
+
+    let agg_ctx = dzkp_validator.context();
+
+    let agg_work = stream::iter(report_pairs)
+        .enumerate()
+        .map(|(idx, reports)| {
+            let agg_ctx = agg_ctx.clone();
+            async move {
+                let (breakdown_key, _) = integer_add::<_, EightBitStep, 1>(
+                    agg_ctx.narrow(&AggregateReportsStep::AddBK),
+                    idx.into(),
+                    &reports[0].breakdown_key.to_bits(),
+                    &reports[1].breakdown_key.to_bits(),
+                )
+                .await?;
+                let (value, _) = integer_add::<_, EightBitStep, 1>(
+                    agg_ctx.narrow(&AggregateReportsStep::AddV),
+                    idx.into(),
+                    &reports[0].value.to_bits(),
+                    &reports[1].value.to_bits(),
+                )
+                .await?;
+                Ok::<_, Error>(AggregateableHybridReport::<BK, V> {
+                    match_key: (),
+                    breakdown_key: breakdown_key.collect_bits(),
+                    value: value.collect_bits(),
+                })
+            }
+        });
+
+    validated_seq_join(dzkp_validator, agg_work)
+        .try_collect()
+        .await
```

Comment:
cc @akoshelev this seems to work

Reply:
oh well...send bounds issue is really nasty

Reply:
maybe I'm misunderstanding. isn't this better, since we no longer use boxed?